### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by andysaputra**

### CODE REVIEW

File: examples/master-with-read-replica/main.tf [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  service_name   = "beitest"
+  environment    = "special"
+  description    = "Master DB"
+
+  instance_class = "db.t2.small"
+  engine_version = "9.6.6"
+
+  allocated_storage = 1024
+
+  multi_az = false
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: true

File: examples/master-with-read-replica/main.tf [hunk 2]
```diff
@@ -0,0 +1,73 @@
+provider "aws" {
+  region = "ap-southeast-1"
+}
+
+module "master" {
+  source = "../../"
+
+  product_domain = "bei"
+  service_name   = "beitest"
+  environment    = "special"
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: production

File: examples/master-with-read-replica/main.tf [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    "sg-50036436"
+  ]
+
+  # Change to valid db subnet group nam
+  db_subnet_group_name = "tvlk-dev-rds-subnet-group"
+
+  # Change to valid parameter group name
+  parameter_group_name = "default.postgres9.6"
+
+  apply_immediately = true
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: false

File: examples/master-with-read-replica/main.tf [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  maintenance_window      = "Mon:00:00-Mon:03:00"
+  backup_retention_period = 1
+
+  skip_final_snapshot = true
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: false

File: examples/master-with-read-replica/main.tf [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+  # Change to valid db subnet group nam
+  db_subnet_group_name = "tvlk-dev-rds-subnet-group"
+
+  # Change to valid parameter group name
+  parameter_group_name = "default.postgres9.6"
+
+  apply_immediately = true
+  maintenance_window      = "Mon:00:00-Mon:03:00"
+  backup_retention_period = 1
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: 7

File: examples/master-with-read-replica/main.tf [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+  skip_final_snapshot = true
+
+  # Change to valid monitoring role arn
+  monitoring_role_arn = "arn:aws:iam::517530806209:role/rds-monitoring-role"
+
+  # Change to valid route 53 zone id
+  route53_zone_id = "Z32OEBZ2VZHSJZ"
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: let's remove r53 record from this module

File: examples/master-with-read-replica/main.tf [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+module "read_replica" {
+  source = "../../modules/read-replica"
+
+  product_domain = "bei"
+  service_name   = "beitest"
+  environment    = "special"
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: production

File: examples/master-with-read-replica/main.tf [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  replicate_source_db = "${module.master.id}"
+
+  instance_class = "db.t2.small"
+
+  # Change to valid security group id
+  vpc_security_group_ids = [
+    "sg-50036436"
+  ]
+
+  availability_zone = "${module.master.availability_zone}"
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: bad example. What if master is multi-az?

File: examples/master-with-read-replica/main.tf [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+  # Change to valid parameter group name
+  parameter_group_name = "default.postgres9.6"
+
+  maintenance_window = "Mon:00:00-Mon:03:00"
+
+  # Change to valid monitoring role arn
+  monitoring_role_arn = "arn:aws:iam::517530806209:role/rds-monitoring-role"
+
+  # Change to valid route 53 zone id
+  route53_zone_id = "Z32OEBZ2VZHSJZ"
```
Comment (sshvetsov) [inline examples/master-with-read-replica/main.tf]: let's remove r53 record form this module

File: variables.tf [hunk 10]
```diff
@@ -0,0 +1,142 @@
+variable "product_domain" {
+  description = "The name of the product domain this RDS belongs to"
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: variables.tf [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+}
+
+variable "service_name" {
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: variables.tf [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+  description = "The name of the service this RDS belongs to, this will be part of the database identifier"
+}
+
+variable "environment" {
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: variables.tf [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+  description = "The environment this RDS belongs to"
+}
+
+variable "description" {
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: variables.tf [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+  description = "The description of this RDS instance"
+}
+
+variable "engine_version" {
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: variables.tf [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+  description = "The postgres engine version"
+}
+
+variable "instance_class" {
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: variables.tf [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+  description = "The instance type of the RDS instance"
+}
+
+variable "username" {
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: variables.tf [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+  description = "Username for the master DB user"
+  default     = "postgres"
+}
+
+variable "password" {
```
Comment (sshvetsov) [inline variables.tf]: add explicit type

File: main.tf [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+resource "aws_db_instance" "postgres" {
+  identifier     = "${local.identifier}"
+
+  engine         = "postgres"
+  engine_version = "${var.engine_version}"
+  instance_class = "${var.instance_class}"
+  username       = "${var.username}"
+  password       = "${var.password}"
+  name           = "${var.name}"
```
Comment (sshvetsov) [inline main.tf]: I don't think this should be in the module. We don't typically create databases via Terraform.

File: variables.tf [hunk 19]
```diff
@@ -0,0 +1,142 @@
...
+  description = "Password for the master DB user. Note that this may show up in logs, and it will be stored in the state file"
+  default     = "changemeplease"
+}
+
+variable "name" {
```
Comment (sshvetsov) [inline variables.tf]: Let's remove this, since we don't normally create DBs using Terraform

File: variables.tf [hunk 20]
```diff
@@ -0,0 +1,142 @@
+variable "product_domain" {
```
Comment (sshvetsov) [inline variables.tf]: Add explicit type to all variables

File: variables.tf [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+variable "monitoring_interval" {
+  description = "The interval, in seconds, between points when Enhanced Monitoring metrics are collected for the DB instance"
+  default     = 60
+}
+
+variable "monitoring_role_arn" {
+  description = "The ARN for the IAM role that permits RDS to send enhanced monitoring metrics to CloudWatch Logs"
+}
+
+variable "route53_zone_id" {
```
Comment (sshvetsov) [inline variables.tf]: Let's remove this and create records outside of the module.

File: main.tf [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+  identifier = "${var.service_name}-postgres-${random_id.suffix.hex}"
+
+  final_snapshot_identifier = "${local.identifier}-final-snapshot"
+}
+
+resource "random_id" "suffix" {
+  byte_length = 8
+}
+
+resource "aws_db_instance" "postgres" {
```
Comment (sshvetsov) [inline main.tf]: rename to "this"

File: outputs.tf [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+  description = "The RDS Resource ID of this instance"
+  value       = "${aws_db_instance.postgres.resource_id}"
+}
+
+output "availability_zone" {
+  description = "The availability zone of the instance"
+  value       = "${aws_db_instance.postgres.availability_zone}"
+}
+
+output "route53_fqdn" {
```
Comment (sshvetsov) [inline outputs.tf]: Let's remove

File: main.tf [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+
+  allocated_storage = "${var.allocated_storage}"
+  storage_type      = "${var.storage_type}"
+  iops              = "${var.iops}"
+  storage_encrypted = "${var.storage_encrypted}"
+  kms_key_id        = "${var.kms_key_id}"
+
+  vpc_security_group_ids = ["${var.vpc_security_group_ids}"]
+  multi_az               = "${var.multi_az}"
+  publicly_accessible    = false
```
Comment (syhrz) [inline main.tf]: shall we make it as variable with default as false?

**Comment by rizkidoank** [comment]
I didn't found any issues other than other reviewers comment. Did you have tried modifying the existing infra created by this module?
Did we have a case that using different parameter group and security group for read replica. If yes, it was multiple apply required, because read replica will use same parameter with the master. I don't know what is the current behaviour though. If this still happens, I think its nice to mention it in the readme.

**Comment by andysaputra** [comment]
Yes, I have tried creating master with read replica, then promote the read replica to be a dedicated instance. The steps are already listed on the README.md.

I see, you are correct. Read replica will use the same parameter group as the master's. I guess for now I will just mention it on the README.md that read replica's parameter group must always be the same as the master's (the same as allocated_storage), because if we leave it as default, there's a chance the resulted state is wrong.

**Comment by rizkidoank** [comment]
Here is some parameters that I found will inherit from master
- parameter group
- security group
- maintenance window
Not sure if there is any other params, better look into the docs or provider.

Sure, thanks. other than that seems looks good to me.

File: main.tf (hunk 24 above)
Comment (sshvetsov) [reply main.tf]: No. There's no good reason for RDS to be public (other than laziness or sloppyness and disregard for security).

**Pull request merged**